networkx/algorithms/find_euler_path.py

```diff
@@ -0,0 +1,223 @@
+    
+        if g.is_directed():
+            degree = g.in_degree
+            out_degree = g.out_degree
+            edges = g.in_edges_iter
+            get_vertex = itemgetter(0)
+            directed = True
+        else:
+            degree = g.degree
+            edges = g.edges_iter
+            get_vertex = itemgetter(1)
+            
+        # Verify if an euler path can be found. Complexity O(n) ?
+        for vertex in g.nodes():
+            deg = degree(vertex)
+            # directed case
+            if directed:
+                outdeg = out_degree(vertex)
+                if deg != outdeg:
+                    # if we have more than 2 odd nodes, we do a raise (no euler path)
+                    if len(check_odd) > 2:
+                        raise nx.NetworkXError("G doesn't have an Euler Path.")
+                # is odd and we append it. 
+                    else:
+                        check_odd.append(vertex)
+            # undirected case        
+            else:
+                if deg % 2 != 0:
+                    # if we have more than 2 odd nodes, we do a raise (no euler path)
+                    if len(check_odd) > 2:
+                        raise nx.NetworkXError("G doesn't have an Euler Path.")
+                    # is odd and we append it. 
+                    else:
+                        check_odd.append(vertex)
+
+        if directed:
+            def verify_odd_cond(g,check_odd):
+                first = check_odd[0]
+                second = check_odd[1]
+                if  g.out_degree(first) == g.in_degree(first) + 1 and g.in_degree(second) == g.out_degree(second) + 1:
+                    return second
+                elif g.out_degree(second) == g.in_degree(second) + 1 and g.in_degree(first)  == g.out_degree(first) + 1:
+                    return first
+                else:
+                    return None
+            start = verify_odd_cond(g,check_odd)
+        else:
+            start = check_odd[0]
+
+        # if the odd condition is not meet, raise an error.
+        if not start:
+            raise nx.NetworkXError("G doesn't have an Euler Path")
+        # Begin algorithm: 
+        vertex_stack = [start]
+        last_vertex = None
+
+        while vertex_stack:
+
+            current_vertex = vertex_stack[-1] #(4)
+            # if no neighbors:
+            if degree(current_vertex) == 0:
+                # Special case, we cannot add a None vertex to the path.
+                if last_vertex is not None:
+                    yield (last_vertex, current_vertex)
+                last_vertex = current_vertex
+                vertex_stack.pop()
+            # we have neighbors, so add the vertex to the stack (2), take any of its neighbors (1)
+            # remove the edge between selected neighbor and that vertex,
+            # and set that neighbor as the current vertex (4).
+            else:
+                random_edge = next(edges(current_vertex)) #(1)
```

Comment:
This is not a "random" edge, it is an "arbitrary" edge. Better to just name the variable `edge`.